packages/flutter/lib/src/material/page_transitions_theme.dart

```diff
@@ -672,13 +779,204 @@ class ZoomPageTransitionsBuilder extends PageTransitionsBuilder {
+      enabledCallback: () => route.popGestureEnabled,
+      builder: (BuildContext context, AndroidBackEvent? startBackEvent,
+          AndroidBackEvent? currentBackEvent) {
+        final bool linearTransition = PageRoute.isPopGestureInProgress(route);
+
+        if (linearTransition) {
+          return backGestureTransitionBuilder(
+            route,
+            context,
+            animation,
+            secondaryAnimation,
+            child,
+            startBackEvent,
+            currentBackEvent,
+          );
+        }
+
+        return parent.buildTransitions(
+          route,
+          context,
+          animation,
+          secondaryAnimation,
+          child,
+        );
+      },
+    );
+  }
+
+  static Widget defaultBackGestureTransitionBuilder<T>(
+    PageRoute<T> route,
+    BuildContext context,
+    Animation<double> animation,
+    Animation<double> secondaryAnimation,
+    Widget child,
+    AndroidBackEvent? startBackEvent,
+    AndroidBackEvent? currentBackEvent,
+  ) {
+    return AndroidBackGestureTransition(
+      animation: animation,
+      secondaryAnimation: secondaryAnimation,
+      getIsCurrent: () => route.isCurrent,
+      child: child,
+    );
+  }
+}
+
+typedef AndroidBackGestureDetectorWidgetBuilder = Widget Function(
+  BuildContext context,
+  AndroidBackEvent? startBackEvent,
+  AndroidBackEvent? currentBackEvent,
+);
+
+class AndroidBackGestureDetector extends StatefulWidget {
+  const AndroidBackGestureDetector({
+    super.key,
+    required this.enabledCallback,
+    required this.backGestureController,
+    required this.builder,
+  });
+
+  final ValueGetter<bool> enabledCallback;
+  final GestureTransitionController backGestureController;
+  final AndroidBackGestureDetectorWidgetBuilder builder;
+
+  @override
+  State<AndroidBackGestureDetector> createState() =>
+      _AndroidBackGestureDetectorState();
+}
+
+class _AndroidBackGestureDetectorState extends State<AndroidBackGestureDetector>
+    with WidgetsBindingObserver {
+  AndroidBackEvent? _startBackEvent;
+
+  AndroidBackEvent? get startBackEvent => _startBackEvent;
+
+  set startBackEvent(AndroidBackEvent? startBackEvent) {
+    if (_startBackEvent != startBackEvent && mounted) {
+      _startBackEvent = startBackEvent;
+      setState(() {});
+    }
+  }
+
+  AndroidBackEvent? _currentBackEvent;
+
+  AndroidBackEvent? get currentBackEvent => _currentBackEvent;
+
+  set currentBackEvent(AndroidBackEvent? currentBackEvent) {
+    if (_currentBackEvent != currentBackEvent && mounted) {
+      _currentBackEvent = currentBackEvent;
+      setState(() {});
+    }
+  }
+
+  bool gestureInProgress = false;
+
+  @override
+  void initState() {
+    super.initState();
+    WidgetsBinding.instance.addObserver(this);
+  }
+
+  @override
+  void dispose() {
+    WidgetsBinding.instance.removeObserver(this);
+    super.dispose();
+  }
+
+  @override
+  Future<bool> startBackGesture(AndroidBackEvent backEvent) {
```

Comment:
This, update, and cancel get called when I tap on the back button in the AppBar. That's what's causing the weird behavior that I mentioned when tapping the back button.

Let me know if you still can't reproduce this.